src/features/courses/course-controller.ts

```diff
@@ -2522,6 +2654,335 @@ class CourseController {
+        } else if (topic.topicTypeId === 2) {
+            let topicIsLive = false;
+            if (_.isNil(studentTopicAssessmentInfoId)) {
+                // specific version was not supplied - see if there's a live version for this question
+                const gradeInstance = await courseRepository.getCurrentInstanceForQuestion({questionId, userId: user.id});
+                topicIsLive = !_.isNil(gradeInstance); // if we got a current instance, then topic is live
+            } else {
+                const studentTopicInfo = await this.getStudentTopicAssessmentInfoById(studentTopicAssessmentInfoId);
+                topicIsLive = studentTopicInfo.isClosed === false && // isClosed might not be accurate when the assessment times out
+                    moment().isBetween(studentTopicInfo.startTime.toMoment(), studentTopicInfo.endTime.toMoment());
+            }
+            const topicInfo = await topic.getTopicAssessmentInfo();
+            if (topicIsLive) {
+                return true;
+            } else {
+                // strike that -- reverse it ;)
+                return !topicInfo.hideProblemsAfterFinish;
+            }
+        }
+        // we *should* never get here - but if we do, then the answer is NO
+        return false;
+    };
+
+    /*
+    * Determine if a user (student) is allowed to start a new version
+    * No one may start a new version if one is already open
+    * If the question belongs to an assessment, check for a current version unless a versionId is provided
+    */
+    async canUserStartNewVersion(options: UserCanStartNewVersionOptions): Promise<UserCanStartNewVersionResult> {
+        const {user, topicId} = options;
+        let userCanStartNewVersion = true;
+        let message = '';
+
+        let topic = await this.getTopicById({ id: topicId }); // includes TopicOverrides
+        const topicInfo = await this.getTopicAssessmentInfoByTopicId({
+            topicId: topicId,
+            userId: user.id
+        }); // method result is ordered by startDate, includes overrides
+        const versions = await this.getStudentTopicAssessmentInfo({
+            topicAssessmentInfoId: topicInfo.id,
+            userId: user.id
+        });
+
+        // deal with overrides
+        const maxVersions = topicInfo.studentTopicAssessmentOverride?.[0]?.maxVersions ?? topicInfo.maxVersions;
+        if (!_.isNil(topic.studentTopicOverride)) {
+            topic = topic.getWithOverrides(topic.studentTopicOverride[0]) as CourseTopicContent;
+        }
+
+        // restrictions on start time, maximum versions, and version delay only apply to students
+        if (user.roleId === Role.STUDENT) {
+            // if the assessment has not yet started, students cannot generate a version
+            if (new Date().getTime() < topic.startDate.getTime()) {
+                message = `The topic "${topic.name}" has not started yet.`;
+                userCanStartNewVersion = false;
+            } else
```

Comment:
I think we should inline `else if`s together. Personally, it makes it harder for me to see them as a contiguous block- I initially read these as noops.

Reply:
💯 (some reactions aren't available to attach)

Reply:
[Noops](https://i.pinimg.com/originals/0b/51/6e/0b516e1113ca5f8ebb094af23e47a5ce.jpg)